Add Integrator Anti-Windup Article

noguchi-takahiro opened the pull request:
This is to close #142.

I made a draft of an integrator anti-windup section in docs.AMDC, especially [control-with-amdc](https://github.com/Severson-Group/docs.amdc.dev/tree/takahiro_docs/source/getting-started/control-with-amdc).

The article draft is available [anti-windup_draft.pdf](https://github.com/Severson-Group/docs.amdc.dev/files/12338534/anti-windup_draft.pdf) and  source of the article is [here](https://github.com/Severson-Group/docs.amdc.dev/tree/takahiro_docs/source/getting-started/control-with-amdc/integrator-anti-windup).

The contents include:

- How to think about performance of anti-windup
- Different methods
    - clamping(continuous)
    - clamping(discrete-time)
    - back-tracking(continuous)
    - back-tracking(discrete-time)

Please review this and feel free to give me feedback. Major and minor changes or proofreading are welcome.

npetersen2 committed:
Add initial template of new docs control content per #77

noguchi-takahiro committed:
Add images

noguchi-takahiro committed:
Add scripts

noguchi-takahiro committed:
Modify a layout

elsevers reviewed (changes requested):
Thank you @noguchi-takahiro! Nice job -- and I can see a lot of work went into this.  I compared notes with @npetersen2 and typed up some feedback on your article here. Our consensus is that this is close, but we want to adjust the perspective a bit, which will have you remove some content and then adjust your structure.

Feedback:
1. Framing the article: The main purpose is explaining to people that you need to simulate realistic scenarios of windup and antiwindup performance because there is not a perfect solution. It depends on how long you are in windup, how extreme it is, etc.
    - Can’t really have perfect anti-windup, need to have simulations  to investigate specific scenarios that you expect to encounter
4. Pick an abstract, simple first order plant to use as an example as you develop the techniques
    - Plant: 1/(s+1), Controller: PI (kp + ki/s) (pick a tuning to give a first order response with a bandwidth of 10 Hz).
    - Comment on how this example could be viewed as a voltage (input) and current (output) for an RL load, torque (input) and speed (output), or heat (input) and temperature (output)
2. It is nice that you show a universal block diagram, with an anti-windup block that gets populated with different contents for clamping and back stepping. Let's have you revise those clamping and back stepping diagrams to make it more clear that they are showing the contents of this block. We suggest to put a box around these with input and output ports clearly labeled that match your universal diagram ![image](https://github.com/Severson-Group/docs.amdc.dev/assets/7421348/06119a9d-5367-437f-b465-4c20a1ccec08)
2. Only discuss and develop continuous time techniques. We will plan to add the discrete time implementation later, as part `Application` articles, such as the planned discrete time CVCR article.
5. As the article progresses, take the following structure:
    1. demonstrate the PI controller and plant with no anti-windup and show the types issues of issues that an engineer will encounter
        - discuss and show example with both **command** and **disturbance** induced anti-windup
    2. Then introduce a simple version of clamping that simply stops integrating when preSat != postSat. 
        - Show plots with multiple traces: original and simple clamped response, again for both **command** and **disturbance** induced anti-windup
    3. Then introduce an advanced version of clamping *(the block diagram and description that you currently are showing)*. 
        - Show plots with multiple traces: original, simple clamped response, advanced clamped
        - Make sure you have a case in your example that showcases the advanced version working better and comment on that (this needs to be sufficiently compelling to motivate your reader to go through the effort to implement the more advanced version)
    4. Then introduce back stepping
        - Show plots with multiple traces: original, advanced clamped response, back stepping
        - Discuss how you pick Kb. The choice of picking Kb is very nuanced and depends on the type of event that you are handling. If you make a poor choice for Kb, clamping will actually be better. 
            - This article should help the reader understand this nuance and that in some cases clamping is a better choice.
            - Show a results plot for where you come up with a Kb value that works really well for the example system + command tracking/disturbances. Now, keep the same plant and same controller and Kb value, but change your command or disturbances so that the backtracking now yields inferior results as compared to clamping. 

Thanks @noguchi-takahiro!! I'm happy to set up a meeting to discuss if that would be helpful.

noguchi-takahiro commented:
Thank you for your feedback @elsevers @npetersen2 . I have a question about item 5,

> discuss and show example with both command and disturbance induced anti-windup

Does the disturbance here mean that I need to add the disturbance voltage between the controller and plant?

elsevers commented:
> Thank you for your feedback @elsevers @npetersen2 . I have a question about item 5,
> 
> > discuss and show example with both command and disturbance induced anti-windup
> 
> Does the disturbance here mean that I need to add the disturbance voltage between the controller and plant?

The answer is yes, but this brings up an important point. So it is confusing when you think about this in terms of current regulation, right? Like... why is there a disturbance voltage? What is the physical basis for this? 

This is part of the reason for point 2:

> Pick an abstract, simple first order plant to use as an example as you develop the techniques
> ...
> Comment on how this example could be viewed as a voltage (input) and current (output) for an RL load, torque (input) and speed (output), or heat (input) and temperature (output)

So we're thinking you should just have this be an abstract first order system, that you occasionally relate back to a real system. If this first order system represents torque and speed, you can more easily imagine how you'd have a torque disturbance from your load, right?

noguchi-takahiro commented:
Thank you @elsevers. I am trying to make the abstract, simple Simulink model based your comment. I have some updates and question about the item 2 and the simulation results.

> 2. Pick an abstract, simple first order plant to use as an example as you develop the techniques
Plant: 1/(s+1), Controller: PI (kp + ki/s) (pick a tuning to give a first order response with a bandwidth of 10 Hz).

Here is the updated Simulink model that demonstrates the simple example;
![image](https://github.com/Severson-Group/docs.amdc.dev/assets/114006024/2299647e-5e48-4527-95c9-d5896c6918ac)

The plant model has a transfer function of $G_p =1/(s+1)$, and the PI controller is tuned as $K_P=\omega_b$ and $K_I=\omega_b$, where $\omega_b=2 \pi \times10$. For now, I am implementing the simple version of clamping as the anti-windup method, where the saturation values are set as $\pm 10$.

Here are results of the responses.

1.) Without anti-windup
![image](https://github.com/Severson-Group/docs.amdc.dev/assets/114006024/c6871d8d-c017-444e-b1a9-966a0d6a99a0)

2.) With anti-windup
![image](https://github.com/Severson-Group/docs.amdc.dev/assets/114006024/411c1f7e-8244-42ac-80ea-db7fe7999d2f)

where the command is set as 1 at the time of 0.2 sec. You can see the overshoot is improved in the case of the with anti-windup. I think these results are reasonable for me.

Next, I wanted to try inject the disturbance of a pulse to see what would happen. The disturbance pulse has the amplitude of 1 during 0.1 second as shown in the figure below. 

![image](https://github.com/Severson-Group/docs.amdc.dev/assets/114006024/3432f4a4-d6b0-4afc-a8bc-65c8b54121dd)
 
where the command is set as 0 for this simulation. Note that the simple clamping is activated in this simulation. However, the anti-windup is not so effective this time as the manipulated value is no longer saturated. If we want to see the effectiveness of the anti-windup with the disturbance, could I just increase the disturbance so that the windup occurs? Is there any robust simulation scenario that can help determine the relationship between disturbance and anti-windup?

Or is this the right way that I should take here in the first place?

elsevers commented:
Thanks @noguchi-takahiro. It looks like you are on the right track here. In regards to your question, I think you do want to increase your disturbance so that your regulator saturates. 

@npetersen2, please chime in if you have further thoughts on this.

noguchi-takahiro commented:
Thank you your comments @elsevers and @npetersen2!

I got two questions this time about the item 5-iii and 5-iv

# addressing item 5-iii

<p align="center">
<img width="600" alt="image" src="https://github.com/Severson-Group/docs.amdc.dev/assets/114006024/30cb2122-601f-4844-9017-5beaccd1a72e">
</p>

As I understand, I need to come up with a simulation scenario in which the advanced clamping produces a better result. Here is an example simulation I implemented this time, where the command of 1 is issued at 0.2 sec, and of -4 is issued at 1 sec.

<p align="center">
<img width="600" alt="image" src="https://github.com/Severson-Group/docs.amdc.dev/assets/114006024/5413f60e-d153-4a63-a792-5be22ebe0671">
</p>

The figures below are expanded versions around the time 0.2 and 1.5 sec, respectively.
(a) 0.2 sec

<p align="center">
<img width="600" alt="image" src="https://github.com/Severson-Group/docs.amdc.dev/assets/114006024/137fa4b3-7b12-4e88-9702-870ab79be2de">
</p>

(a) 1.5 sec

<p align="center">
<img width="600" alt="image" src="https://github.com/Severson-Group/docs.amdc.dev/assets/114006024/767895ee-c420-4116-affa-c547be1512ba">
</p>

It looks like the simple clamping (blue) and the advanced clamping (dashed cyan) are a complete match. But we want to separate them, i.e., get the advanced clamping better... 

When the process of the advanced clamping was done, the integration was stopped only if BOTH conditions below were satisfied.

1. The preSat (= controller output) is saturated
2.  The sign of error and the preSat are not equal

I think if condition 1 is satisfied and the I-controller stops the integration, it is equivalent to simple clamping. So, I needed to come up with a situation that satisfies condition 2 to compare the simple and the advanced clamping. My question is, when are the sign of error and the present not equal? I tried several scenarios, changing the command waveforms, but these methods were equal, i.e., I could not find a nice scenario and compare them. For your information, I referred to this advanced clamping method from these resources:

- [PID Controllers , Theory , Design and Tuning (2nd Edition)](https://aiecp.wordpress.com/wp-content/uploads/2012/07/1-0-1-k-j-astrom-pid-controllers-theory-design-and-tuning-2ed.pdf) to page 88.

<p align="center">
<img width="600" alt="image" src="https://github.com/Severson-Group/docs.amdc.dev/assets/114006024/3914a586-7093-45b1-905d-c527e947535f">
</p>


- [Integral Anti-Windup for PI Controllers](https://www.researchgate.net/profile/Mohamed-Mourad-Lafifi/post/What-does-it-mean-that-integrator-creates-saturation-and-the-control-theory-using-an-integrator-that-saturates-becomes-invalid/attachment/59d648c879197b80779a3712/AS%3A468515938410497%401488714312270/download/Integral+Anti-Windup+for+PI+Controllers.pdf)
- [Anti-windup for PID control | Understanding PID Control, Part 2](https://www.mathworks.com/videos/understanding-pid-control-part-2-expanding-beyond-a-simple-integral-1528310418260.html)

# addressing item 5-iv

<p align="center">
<img width="600" alt="image" src="https://github.com/Severson-Group/docs.amdc.dev/assets/114006024/811a5ab6-23f8-4eee-8377-d6e054bb8b38">
</p>

This is similar to the previous question, but how can we come up with the scenario that the clamping method is better than the back-tracking by just modifying the command/disturbance? If I command the issue of 1 at 0.2 sec, it looks like the back-tracking method is better than clamping (see the previous results), whereas if I inject the disturbance of 15 for 0.1 sec, the clamping is better around the transient response, as shown below. I guess this is not always true, depending on the back-tracking gain of $K_b$, but how can we make these phenomena more compelling? Notice that $K_b=K_I/K_P=1$.

<p align="center">
<img width="600" alt="image" src="https://github.com/Severson-Group/docs.amdc.dev/assets/114006024/7f4293fa-4266-484f-9f28-5a78aa81bdfe">
</p>

<p align="center">
<img width="600" alt="image" src="https://github.com/Severson-Group/docs.amdc.dev/assets/114006024/8cb67db3-0b8c-4f31-a4c1-eb3dbb11dd0a">
</p>

I would be glad if you could give me your thought here.

npetersen2 commented:
Good discussion and thoughts @noguchi-takahiro , interesting.

I have a feeling that the scenario which will help show the differences in these methods for anti-windup is a "fast" transient from positive windup, to negative windup, back to positive. Let me give an example:

Imagine a position tracking system holding the position to 0. The max control effort is saturated to 10. Then, a user applies a +15 disturbance for 1 second, then immediately applies a -15 disturbance for 0.5 sec, then back to +15, then 0. I think this "slamming" of the system will help show the differences in the anti-windup approaches.

What do you think?

noguchi-takahiro committed:
Add Simulink file

noguchi-takahiro commented:
Thank you for your feedback @npetersen2 @elsevers .

I am still trying to address your comment, but could you check if my Simulink model is properly implemented when you have a time?

I temporary added the MATLAB script [main.m](https://github.com/Severson-Group/docs.amdc.dev/blob/630892303b5ecac740df5096f93a329dd6667988/source/getting-started/control-with-amdc/integrator-anti-windup/main.m) to run the [Simulink file](https://github.com/Severson-Group/docs.amdc.dev/blob/630892303b5ecac740df5096f93a329dd6667988/source/getting-started/control-with-amdc/integrator-anti-windup/simple_model.slx).

npetersen2 commented:
Thanks @noguchi-takahiro , I will try to take a look at this soon.

npetersen2 commented:
Also, a very interesting approach for anti-wind-up is presented in this article: https://ieeexplore.ieee.org/document/8279496

I do not suggest you add this to your article right now, but eventually, it would be nice to explore this approach and use it. It looks quite nice and useful!

npetersen2 commented:
Hey @noguchi-takahiro , I see this Simulink model was created in `R2023b`.

Unfortunately, I only have `R2023a` installed... would you be able to open your model and save-as `R2023a` format? Then I can easily take a look

noguchi-takahiro committed:
Replace R2023b with R2023a

noguchi-takahiro commented:
Thank you @npetersen2. I hope the Simulink model was saved as the earlier version `R2023a`. Could you check if this [simple_model.slx](https://github.com/Severson-Group/docs.amdc.dev/blob/9e22b7cf96f835f07aa54d2672d4487074b28065/source/getting-started/control-with-amdc/integrator-anti-windup/simple_model.slx) is combatable with R2023a?

npetersen2 commented:
I am commenting here to archive some Slack conversations I have had with @noguchi-takahiro about this:

------

I have been investigating the anti-windup methods via your simulation, specifically the "advanced" and the "simple" clamping methods

I believe you found they behaved the same way... so no reason to use the advanced method when the simple method works well

I finally managed to find a contrived example when they do NOT behave the same, where I believe the advanced method works better

![image](https://github.com/Severson-Group/docs.amdc.dev/assets/20168990/45af631e-6df4-434a-8209-b3970336032d)

- In the beginning, there is a large disturbance which causes the system to saturate, and it steady-states at -1 output....
- Then, at 5 sec, the controller set point starts toggling around this disturbance driven output point

the first time the controller set point goes below the output point, it triggers the advanced method to NOT clamp the integrator, while the simple method stays clamped. This causes the advanced method to un-windup in a good way, where the simple one cannot. So, from this point onward, the advanced method behaves better

HOWEVER

this only occurs if the `Kp` gain is basically not present... i.e. really small.

for typical `Kp` tuning, at this instance in the simulation, the proportional control output causes the output to drop out of saturation, which causes there to be no differences in the advanced vs simple methods


so, my summary:

- the advanced method does actually make sense and is logical
- however, for our typical systems, it is quite unlikely to ever actually result in different behavior compared to the simple method
- the only case of difference is when: we are in saturation AND the control output would be driving us out of saturation (i.e. signs are different from error vs control output preset),
   -this is quite rare, since in this case, usually the difference in signs immediately moves us OUT of saturation
   - so, for this to matter, we need an example where we can stay in saturation during this flip in signs
   - so far, i can only find an example of this by basically removing the Kp gain


if you run this, it should output the figure above: [sim_files.zip](https://github.com/user-attachments/files/15537486/sim_files.zip)

noguchi-takahiro committed:
Update MATLAB/Simulin model

noguchi-takahiro committed:
Update document

noguchi-takahiro committed:
Update documents and figures

noguchi-takahiro committed:
Update document

noguchi-takahiro commented:
@elsevers @npetersen2 I believe that I addressed your feedback and ready for getting your additional review.

Here is [source](https://github.com/Severson-Group/docs.amdc.dev/blob/takahiro_docs/source/getting-started/control-with-amdc/integrator-anti-windup/index.md) of the anti-windup article.

This is a rendering version of the article.
[anti-windup_draft_v2.pdf](https://github.com/user-attachments/files/16114527/anti-windup_draft_v2.pdf)

mhmokhtarabadi committed:
Merge branch 'main' into takahiro_docs

mhmokhtarabadi committed:
Add anti-windup article to the control with AMDC page

mhmokhtarabadi committed:
Remove white spaces

noguchi-takahiro committed:
Update source/getting-started/control-with-amdc/integrator-anti-windup/index.md

Co-authored-by: Mohamadhasan <70004596+mhmokhtarabadi@users.noreply.github.com>

noguchi-takahiro committed:
Update source/getting-started/control-with-amdc/integrator-anti-windup/index.md

Co-authored-by: Mohamadhasan <70004596+mhmokhtarabadi@users.noreply.github.com>

noguchi-takahiro committed:
Update source/getting-started/control-with-amdc/integrator-anti-windup/index.md

Co-authored-by: Mohamadhasan <70004596+mhmokhtarabadi@users.noreply.github.com>

noguchi-takahiro committed:
Update source/getting-started/control-with-amdc/integrator-anti-windup/index.md

Co-authored-by: Mohamadhasan <70004596+mhmokhtarabadi@users.noreply.github.com>

noguchi-takahiro committed:
Update source/getting-started/control-with-amdc/integrator-anti-windup/index.md

Co-authored-by: Mohamadhasan <70004596+mhmokhtarabadi@users.noreply.github.com>

noguchi-takahiro committed:
Update source/getting-started/control-with-amdc/integrator-anti-windup/index.md

Co-authored-by: Mohamadhasan <70004596+mhmokhtarabadi@users.noreply.github.com>

noguchi-takahiro committed:
Update source/getting-started/control-with-amdc/integrator-anti-windup/index.md

Co-authored-by: Mohamadhasan <70004596+mhmokhtarabadi@users.noreply.github.com>

noguchi-takahiro committed:
Remove not used folder

noguchi-takahiro committed:
Removed not used folder

noguchi-takahiro committed:
Remove old version file

noguchi-takahiro committed:
Add Gitignore

noguchi-takahiro committed:
Rename all image files

noguchi-takahiro committed:
Create simulink folder

noguchi-takahiro committed:
Rename simulink

noguchi-takahiro committed:
Rename m file

noguchi-takahiro committed:
Remove Disturbance

noguchi-takahiro committed:
Rename disturbance

noguchi-takahiro committed:
Make them non italic

noguchi-takahiro committed:
Fix links to images

noguchi-takahiro committed:
Update readme

noguchi-takahiro committed:
Update readme

noguchi-takahiro committed:
Update image to use nonitalice

noguchi-takahiro committed:
Update readme

noguchi-takahiro committed:
Update readme

noguchi-takahiro committed:
Update readme

noguchi-takahiro committed:
Update advanced image

noguchi-takahiro committed:
Upate figures

noguchi-takahiro committed:
Remove m file

noguchi-takahiro committed:
Add m file

noguchi-takahiro committed:
Update disturbance

noguchi-takahiro committed:
Update image

noguchi-takahiro committed:
Update image

noguchi-takahiro committed:
Update image

noguchi-takahiro committed:
Update readme

noguchi-takahiro committed:
Add error and preSat images

noguchi-takahiro committed:
Updat readme

Daehoon-Sung reviewed (dismissed):
1. Are all files under 300 kB (if not, please carefully assess whether it is worth committing them)? **Yes**
2. Are all files named according to the appropriate [naming convention](https://github.com/Severson-Group/research-repo-template?tab=readme-ov-file#file-naming), i.e., dash-case, camelCase, snake case? **Yes**
3. Do all Markdown files follow the [CONTRIBUTING article template](https://github.com/Severson-Group/.github/blob/main/CONTRIBUTING.md#markdown-documentation-template)? **Yes**
4. Do all links work in the material that the PR is adding? **Yes**
5. Is the PR configured to close the correct issue(s)? **Yes**
6. Did the PR fully address the `Approach` section of the issue(s) it is closing? **Yes**

Please work on addressing any **No** items.

mohamed-dek1 commented:
@elsevers can you review this PR?

Daehoon-Sung commented:
Hello, Professoe @elsevers, @noguchi-takahiro, @mhmokhtarabadi. Can we sync up together during today's office hour, after the lab hour, to complete this article?

noguchi-takahiro commented:
Here is the latest article rendered as PDF as of 2/6/2026: [Integrator Anti-Windup - AMDC Platform.pdf](https://github.com/user-attachments/files/25138295/Integrator.Anti-Windup.-.AMDC.Platform.pdf).

noguchi-takahiro committed:
Update index

noguchi-takahiro committed:
Update figure

mhmokhtarabadi committed:
Merge branch 'main' into takahiro_docs

elsevers committed:
Revise introducion

elsevers committed:
Edit first half of windup intro section

elsevers committed:
Edit  second half of the windup intro section

elsevers committed:
Revise clamping section

elsevers committed:
Revise back tracking

elsevers committed:
Fix typos

elsevers reviewed (changes requested):
Thanks @noguchi-takahiro! This is looking really good.

I have been editting / revising this. Can you please give it a read and make sure you are happy with the contents, that I didn't mess anything up, etc.

A few other requests:

## 1. Fix spacing in first figure:

<img width="251" height="128" alt="Image" src="https://github.com/user-attachments/assets/26723a9e-4ff3-4eee-81a5-ad39491a4f5b" />

## 2. Add missing simulations

Can you add the simulink file for the special case considered for advanced clamping? I ran your simulations and I think this is missing.

## 3. Can we exaggerate the windup impact on disturbance rejection?

Right now, the impact is noticeable, but not all that significant. I am wondering if we were to hold the disturbance longer, would we see a more pronounced effect due to the saturation block? Try holding the disturbance for 0.2s instead of 0.1s, or something similar.

<img width="586" height="154" alt="Image" src="https://github.com/user-attachments/assets/c25cd35e-6ead-45e1-bf17-d5488473d846" />

noguchi-takahiro committed:
Fix spacing in image

noguchi-takahiro committed:
Update simulation

noguchi-takahiro committed:
Update m file to include special parameters for advanced integrator

noguchi-takahiro committed:
Change disturbance range

noguchi-takahiro committed:
Update image

noguchi-takahiro committed:
Update images

noguchi-takahiro committed:
Update index

noguchi-takahiro committed:
Update link

noguchi-takahiro committed:
Update link to image

elsevers committed:
Merge branch 'main' into takahiro_docs

elsevers reviewed (dismissed):
1. Are all images < 40 kB and other files < 300 kB (if not, please carefully assess whether it is worth committing them)? **No**, but it is worth adding them.
2. Are all files named according to the appropriate [naming convention](https://github.com/Severson-Group/research-repo-template?tab=readme-ov-file#file-naming), i.e., dash-case, camelCase, snake case? **Yes**
3. Do all Markdown files follow the [CONTRIBUTING article template](https://github.com/Severson-Group/.github/blob/main/CONTRIBUTING.md#markdown-documentation-template)? **No**, they follow the AMDC docs template
4. Do all links work in the material that the PR is adding? **Yes**
5. Is the PR configured to close the correct issue(s)? **Yes**
6. Did the PR fully address the `Approach` section of the issue(s) it is closing? **Yes**

GREAT. Thanks @noguchi-takahiro
